tutel/experts/ffn.py

```diff
@@ -27,21 +31,49 @@ def update(self, ctx):
         local_experts = ctx.num_local_experts
         self.output_dim = self.output_dim or model_dim
 
-        fc1_weight = torch.empty(1, local_experts, hidden_size, model_dim)
-        fc2_weight = torch.empty(1, local_experts, hidden_size, self.output_dim)
-        fc1_bias = torch.empty(1, local_experts, hidden_size)
-        fc2_bias = torch.empty(1, local_experts, (self.output_dim + ctx.sharded_count - 1) // ctx.sharded_count)
-
-        for i in range(local_experts):
-            fc1 = torch.nn.Linear(model_dim, hidden_size)
-            fc2 = torch.nn.Linear(hidden_size, self.output_dim)
-            fc1_weight[0, i, :, :], fc1_bias[0, i, :] = fc1.weight, fc1.bias
-            fc2_weight[0, i, :, :], fc2_bias[0, i, :] = fc2.weight.t(), fc2.bias[:fc2_bias.size(-1)]
-
-        self.register_parameter(name='batched_fc1_w', param=torch.nn.Parameter(fc1_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc2_w', param=torch.nn.Parameter(fc2_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc1_bias', param=torch.nn.Parameter(fc1_bias.squeeze(0)))
-        self.register_parameter(name='batched_fc2_bias', param=torch.nn.Parameter(fc2_bias.squeeze(0)))
+        fc1_weight = torch.empty(local_experts, hidden_size, model_dim, dtype=dtype, device=device)
+        fc2_weight = torch.empty(local_experts, hidden_size, self.output_dim, dtype=dtype, device=device)
+        fc1_bias = torch.empty(local_experts, hidden_size, dtype=dtype, device=device)
+        fc2_bias = torch.empty(local_experts, (self.output_dim + ctx.sharded_count - 1) // ctx.sharded_count, dtype=dtype, device=device)
+
+        self.batched_fc1_w = torch.nn.Parameter(fc1_weight)
+        self.batched_fc2_w = torch.nn.Parameter(fc2_weight)
+        self.batched_fc1_bias = torch.nn.Parameter(fc1_bias)
+        self.batched_fc2_bias = torch.nn.Parameter(fc2_bias)
+
+        self.reset_parameters()
+
+    def _kaiming_uniform(self, tensor, fan, a=0, nonlinearity='leaky_relu'):
+        gain = init.calculate_gain(nonlinearity, a)
+        std = gain / math.sqrt(fan)
+        bound = math.sqrt(3.0) * std  # Calculate uniform bounds from standard deviation
+        with torch.no_grad():
+            return tensor.uniform_(-bound, bound)
+
+    def reset_parameters(self) -> None:
+        # imitate reset_parameters from torch.nn.Linear
+        mode = 'fan_in'
+        a = math.sqrt(5)
+
+        # init fc1
+        # fan is calculated per expert; index weight for expert [0]; note the transpose
+        fan_in, fan_out = init._calculate_fan_in_and_fan_out(self.batched_fc1_w[0].t())
```

Comment:
Can we do something like
```suggestion
        fan_in, fan_out = self.batched_fc1_w.shape[1:]
```
? I'm scared of calling private funcs in 3rd-party libs, and it would be more readable if I didn't have to know what this function did to understand this line.

(not sure this actually preserves the semantics though since I don't know what the private function does)